plenum/server/replica.py

```diff
@@ -1282,6 +1285,11 @@ def _apply_pre_prepare(self, pre_prepare: PrePrepare, sender: str) -> Optional[i
                 old_state_root,
                 old_txn_root))
 
+        for req_key in pre_prepare.reqIdr:
+            ledger_id, seq_no = self.node.seqNoDB.get(req_key)
```

Comment:
It may affect performance... the question is how badly

Reply:
What about txns being ordered (not yet committed)? So, a malicious Primary may send a batch of 100000 same requests.
Or malicious primary may send batches with the same requests in a row, so that previous batched is not yet ordered.